xdsl/tools/xdsl_run.py

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+from xdsl.dialects.builtin import IndexType, StringAttr
+from xdsl.dialects.func import FuncOp
+from xdsl.interpreter import Interpreter
+from xdsl.runner.arith import ArithFunctions
+from xdsl.runner.builtin import BuiltinFunctions
+from xdsl.runner.func import FuncFunctions
+from xdsl.runner.print import PrintFunctions
+from xdsl.runner.scf import ScfFunctions
+from xdsl.utils.hints import isa
+from xdsl.xdsl_opt_main import xDSLOptMain
+
+
+class xDSLRunMain(xDSLOptMain):
```

Comment:
I like the core idea of reusing some of what is available in xDSLOptMain, but I am uncertain it actually makes sense here. Notice that this gives your runner the `pipeline` field for example, which I don't think should be here, or all the command line arguments of xdsl-opt.

AFAIU, you are mainly using this to reuse parts of xdsl opt, which I think makes sense. However, I think it would make more sense to restructure xDSLOptMain to xDSLCLScript, s.t. you can only reuse the parts that actually make sene.

I am fine with having this as an issue for the moment, if you don't want to address it here (which I feel you shouldn't).

Reply:
*totally* agree with this 🙏
Though, my thought was exactly to merge a first iteration to then go from there, because it opens many directions at once.
We can now do different PRs to structure the CLI aspect, improve the interpreter core, the Dialects implementation, and what not

Reply:
I'm personally fine merging it like this, and then making it less hacky later ;)

Reply:
I would recommend against reusing xDSLOptMain here, and I like Michel's suggestion, would vote for splitting this PR into parts and doing the executable class hierarchy before merging